mlos_bench/mlos_bench/dict_templater.py

```diff
@@ -0,0 +1,80 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Simple class to help with nested dictionary $var templating.
+"""
+
+from copy import deepcopy
+from string import Template
+from typing import Any, Dict, Optional
+
+import os
+
+
+class DictTemplater:    # pylint: disable=too-few-public-methods
+    """
+    Simple class to help with nested dictionary $var templating.
+    """
+
+    def __init__(self, source_dict: Dict[str, Any]):
+        """
+        Initialize the templater.
+
+        Parameters
+        ----------
+        source_dict : Dict[str, Any]
+            The template dict to use for source variables.
+        """
+        # A copy of the initial data structure we were given with templates intact.
+        self._template_dict = deepcopy(source_dict)
+        # The source/target dictionary to expand.
+        self._dict: Dict[str, Any] = {}
+
+    def expand_vars(self, *,
+                    extra_source_dict: Optional[Dict[str, Any]] = None,
+                    use_os_env: bool = False) -> Dict[str, Any]:
+        """
+        Expand the template variables in the destination dictionary.
+
+        Parameters
+        ----------
+        extra_source_dict : Dict[str, Any]
+            An optional extra source dictionary to use for expansion.
+        use_os_env : bool
+            Whether to use the os environment variables a final fallback for expansion.
+
+        Returns
+        -------
+        Dict[str, Any]
+            The expanded dictionary.
+        """
+        self._dict = deepcopy(self._template_dict)
+        extra_source_dict = {} if extra_source_dict is None else extra_source_dict
+        self._dict = self._expand_vars(self._dict, extra_source_dict, use_os_env)
```

Comment:
a haskeller in me suggests
```suggestion
        self._dict = self._expand_vars(deepcopy(self._template_dict), extra_source_dict, use_os_env)
```
and make `extra_source_dict` parameter `Optional` in `_expand_vars()`